### ISSUE TITLE
BUG: python2 fails to restore a previously imported module.

### ISSUE THREAD

**Issue by theodoregoetz** [roryyorke/py-hide-modules#1]
I was using this module (thank you by the way) to spare out numpy for testing code that can survive without it, but it doesn't seem to fully restore it in python2 (python3 works as expected). Here is a script that demonstrates the bug:

``` python
from hide_modules import hide_modules

def require_module():
    import numpy
    print('require_module: imported numpy')

def fallback_on_import():
    try:
        import numpy
        print('fallback_on_import: imported numpy')
    except ImportError:
        print('fallback_on_import: could not import numpy')

@hide_modules(['numpy'])
def fn():
    fallback_on_import()

require_module()
fn()
require_module()
```

Expected output (observed with python3):
```
require_module: imported numpy
fallback_on_import: could not import numpy
require_module: imported numpy
```

Observed output with python2:
```
require_module: imported numpy
fallback_on_import: could not import numpy
Traceback (most recent call last):
  File "test-hide-modules.py", line 20, in <module>
    require_module()
  File "test-hide-modules.py", line 4, in require_module
    import numpy
  File "/usr/lib64/python2.7/site-packages/numpy/__init__.py", line 142, in <module>
    from . import add_newdocs
ImportError: cannot import name add_newdocs
```

**Comment by roryyorke** [comment]
Thanks for the report. I'll try to have a look this weekend.

**Comment by roryyorke** [comment]
This error can be reproduced with any package that imports one of its subpackages in __init__.py.

I can get a numpy import error on Python 3 with this simulation of hide_modules current module removal behaviour (also shows effect of sys.modules from your PR):
```python
from __future__ import print_function

import traceback

import numpy
del numpy
import sys
hidden = {'numpy':sys.modules['numpy']}
del sys.modules['numpy']

print('>'*20)

try:
    import numpy
    print('unexpected successful import numpy?')
except ImportError as e:
    print('import numpy: ImportError')
    traceback.print_exc()
except NameError as e:
    print('importing numpy: NameError')
    traceback.print_exc()

print('>'*20)

sys.modules.update(hidden)

import numpy
print('numpy import successful')
```

Python2:
```
$ python2 bug2.py 
>>>>>>>>>>>>>>>>>>>>
import numpy: ImportError
Traceback (most recent call last):
  File "bug2.py", line 14, in <module>
    import numpy
  File "/home/rory/.miniconda3/envs/python2-control-dev/lib/python2.7/site-packages/numpy/__init__.py", line 142, in <module>
    from . import add_newdocs
ImportError: cannot import name add_newdocs
>>>>>>>>>>>>>>>>>>>>
numpy import successful
```

Python 3:
```
$ python3 bug2.py 
>>>>>>>>>>>>>>>>>>>>
importing numpy: NameError
Traceback (most recent call last):
  File "bug2.py", line 14, in <module>
    import numpy
  File "/usr/lib/python3/dist-packages/numpy/__init__.py", line 152, in <module>
    test = testing.nosetester._numpy_tester().test
NameError: name 'testing' is not defined
>>>>>>>>>>>>>>>>>>>>
numpy import successful
```

I don't know (1) why the Python 3 error happens; I guess the import recreates the numpy module, and the fact that numpy.testing is already loaded breaks the setup of the new module; and (2) why this error isn't triggered with current hide_modules.

It seems clear that restoring sys.modules, as in your PR, is a good idea.  An alternative *might* be, when "unloading" a module, to also unload all of its submodules, but as I understand Python doesn't really support module unloading (one can't in general undo the effects of whatever code is run when a module is loaded), so restoring is probably best.

**Comment by roryyorke** [comment]
Fixed by #2.
